xcm/src/v1/multilocation.rs

```diff
@@ -773,9 +827,82 @@ impl TryFrom<MultiLocation> for Junctions {
 #[cfg(test)]
 mod tests {
 	use super::{Ancestor, AncestorThen, Junctions::*, MultiLocation, Parent, ParentThen};
-	use crate::opaque::v1::{Junction::*, NetworkId::Any};
+	use crate::opaque::v1::{Junction::*, NetworkId::*};
 	use parity_scale_codec::{Decode, Encode};
 
+	#[test]
+	fn inverted_works() {
+		let ancestry: MultiLocation = (Parachain(1000), PalletInstance(42)).into();
+		let target = (Parent, PalletInstance(69)).into();
+		let expected = (Parent, PalletInstance(42)).into();
+		let inverted = ancestry.inverted(&target).unwrap();
+		assert_eq!(inverted, expected);
+
+		let ancestry: MultiLocation = (Parachain(1000), PalletInstance(42), GeneralIndex(1)).into();
+		let target = (Parent, Parent, PalletInstance(69), GeneralIndex(2)).into();
+		let expected = (Parent, Parent, PalletInstance(42), GeneralIndex(1)).into();
+		let inverted = ancestry.inverted(&target).unwrap();
+		assert_eq!(inverted, expected);
+	}
+
+	#[test]
+	fn simplify_basic_works() {
+		let mut location: MultiLocation =
+			(Parent, Parent, Parachain(1000), PalletInstance(42), GeneralIndex(69)).into();
+		let context = X2(Parachain(1000), PalletInstance(42));
+		let expected = GeneralIndex(69).into();
+		location.simplify(&context);
+		assert_eq!(location, expected);
+
+		let mut location: MultiLocation = (Parent, PalletInstance(42), GeneralIndex(69)).into();
+		let context = X1(PalletInstance(42));
+		let expected = GeneralIndex(69).into();
+		location.simplify(&context);
+		assert_eq!(location, expected);
+
+		let mut location: MultiLocation = (Parent, PalletInstance(42), GeneralIndex(69)).into();
+		let context = X2(Parachain(1000), PalletInstance(42));
+		let expected = GeneralIndex(69).into();
+		location.simplify(&context);
+		assert_eq!(location, expected);
+
+		let mut location: MultiLocation =
+			(Parent, Parent, Parachain(1000), PalletInstance(42), GeneralIndex(69)).into();
+		let context = X3(OnlyChild, Parachain(1000), PalletInstance(42));
+		let expected = GeneralIndex(69).into();
+		location.simplify(&context);
+		assert_eq!(location, expected);
```

Comment:
Should this pass? Because it does.

```
		let mut location: MultiLocation = (Parent, PalletInstance(42), GeneralIndex(69)).into();
		let context = X2(PalletInstance(30), PalletInstance(42));
		let expected = GeneralIndex(69).into();
		location.simplify(&context);
		assert_eq!(location, expected);
```

I am having a really hard time understanding this function's behavior.

Reply:
Yes. All `location`s within a consensus system must be taken to be within some local `context` (aka the consensus system's _ancestry_). The ancestry is essentially just the consensus system's location from the point of view of the distant ancestor consensus system under which the local consensus sits. For the Polkadot Relay, this is conventionally empty (as it is its own universe) but for a parachain which knows that it exists within a broader consensus system it would be the `Parachain` junction by which the parent Relay-chain refers to it.

Local `location`s which begin with `Parent` should be "escaping" outside of the local consensus system (otherwise there is no need to begin with `Parent`), however if after the `Parent` node(s), the rest of the location is such that it descends back into the local consensus then this becomes a non-canonical location. Basically, the `Parent` items and their corresponding junctions are superfluous because they're already implied by the ancestry.

In order to normalise it, we need to detect and remove this case.

In this case the `context` is `(PalletInstance(30), PalletInstance(42)` 
and the `location` is `Parent, PalletInstance(42), GeneralIndex(69)`. The location is therefore non-canonical, because the `Parent` is being used to descend outside of the local consensus and beyond `PalletInstance(42)` (the last item of `context` and then re-descend immediately back into `PalletInstance(42)`. Therefore this location is exactly equivalent to the simpler (and shorter) `GeneralIndex(69)`.

Reply:
Yes, I understand now. Basically, we use context and the front of the multilocation and cancel out all "redundant" data, and normalize it.

> The ancestry is essentially just the consensus system's location from the point of view of the most distant ancestor consensus system under which the local consensus sits.

This was very key. Thanks.

Reply:
👍 just some note here
![image](https://user-images.githubusercontent.com/1088525/147336679-e5bb8077-f602-49f7-9a55-5ad52f018583.png)

Reply:
where I can read about stuff happening here? except code:) I read all papers on xcm I found and ported acala tests. so found reanchor stuff. do not get what it is. actually security of xcm is very unclear.

Reply:
@dzmitry-lahoda Reanchoring is a piece of logic that simply works behind the scenes. As an end-user of the XCM API, you most likely wouldn't need to touch it. What is the exact problem that led you here to understand reanchoring?

Reply:
I am stealing XCM stuff from Acala.  I am ok with that (0, CurrencyId) => CurrencyId. But here we see that it was sent from Relay. Why it is safe to map (0, CurrencyId) => CurrencyId? May be another buggy para chain may send that and mint infinite amounts? So it is not clear what is security XCM provides to parachains against other parachains.

<img width="1726" alt="image" src="https://user-images.githubusercontent.com/757125/155422460-32079904-ed8f-4cc9-973f-a0e82f426b59.png">

Reply:
@dzmitry-lahoda Substrate/Polkadot doesn't have a concept of `CurrencyId`, and it looks to me to be an Acala/Karura concept. Would you kindly raise this issue in their repo instead?

Reply:
:) Substrate/Polkadot has the location and general key concept. why would somebody trust these are genuine? Asset with location FooBar. Parachain X xcmed FooBar from X to Z. Why would we trust that FooBar was not just minted?

Reply:
I think you are not understanding that every consensus system (or blockchain in layman terms) is responsible for interpreting incoming XCM messages correctly. As such, _how_ a chain decides to interpret `GeneralKey` or `GeneralIndex` is not up to us to decide.

Thus, would you please raise this issue to Acala/Karura?

Reply:
Still not clear. Kusama can have KSM. Kusama can send (if we call its XCM pallet) to send KSM to paracahain. Can someone send KSM back to Kusama? Does Kusama tracks all places it send KSM to be equal?  I mean DOT. Polkadot is relay chain. Does DOT sent via XCM has any security/alignment after it was send from Polkadot to Parachain?

So right now I speak about Relay and some abstract Parachain. And some 2 abstract parachains having DOT/KSM.  How these could interpret correctly message with asset (1, Here) origin?

Reply:
`MultiLocation`s are relative paths, as explained in [Gav's blogpost](https://medium.com/polkadot-network/xcm-the-cross-consensus-message-format-3b77b1373392#1330). A parachain of Kusama/Polkadot will identify KSM/DOT as `(1, Here)`. When they send `(1, Here)` to the relay chain, their local XCM executor will reanchor the location to be `(0, Here)`.

As I mentioned in my first message, reanchoring happens automatically behind the scenes and doesn't require any user input. If you are wondering about how assets are transferred between parachains, I have an [answer on Stack Exchange](https://substrate.stackexchange.com/questions/37/how-can-i-transfer-assets-using-xcm) that describes in detail how it is done, and how the trust issues are resolved.